src/wpublish/QueriesAndChunks.cc

```diff
@@ -489,105 +499,168 @@ QueriesAndChunks::ScanTableSumsMap QueriesAndChunks::_calcScanTableSums() {
+                         __func__ << " a booted uq=" << countQId.qId << ":count=" << countQId.count
+                                  << " uqBootedTasksSum=" << uqBootedTasksSum);
+                }
+            } else {
+                // This really should never happen.
+                LOGS(_log, LOG_LVL_ERROR, __func__ << " queryToBoot is nullptr qIdToBoot=" << countQId.qId);
+            }
+        }
     }
-    _tasks.push_back(task);
-}
 
-/// @return the number of Tasks that have been booted for this user query.
-int QueryStatistics::getTasksBooted() {
-    lock_guard<mutex> guard(_qStatsMtx);
-    return _tasksBooted;
-}
+    // Last check, if any of the queries have several booted `Task`s but have not been booted,
+    // they should be booted now.
+    for (auto const& countQId : countQIdVect) {
+        QueryId qIdToCheck = countQId.qId;
+        // Get query info
+        QueryStatistics::Ptr queryToCheck = getStats(qIdToCheck);
+        if (queryToCheck == nullptr) {
+            LOGS(_log, LOG_LVL_ERROR, __func__ << " Couldn't locate qIdToCheck=" << qIdToCheck);
+            // This really should never happen, but try to boot the next query in the list.
+            continue;
+        }
 
-/// @return true if this query is done and has not been touched for deadTime.
-bool QueryStatistics::isDead(chrono::seconds deadTime, chrono::system_clock::time_point now) {
-    lock_guard<mutex> guard(_qStatsMtx);
-    if (_isMostlyDead()) {
-        if (now - _touched > deadTime) {
-            return true;
+        if (!queryToCheck->_queryBooted) {
+            auto tasksBooted = queryToCheck->getTasksBooted();
+            LOGS(_log, LOG_LVL_DEBUG,
+                 __func__ << " b check uq=" << countQId.qId << " tasksBooted=" << tasksBooted
+                          << " max=" << (_maxTasksBooted));
+            if (tasksBooted > _maxTasksBooted) {
+                QueryStatistics::Ptr queryToBoot = getStats(qIdToCheck);
+                if (_bootUserQuery(queryToBoot, bSched)) {
+                    uqBootedTasksSum += countQId.count;
+                    LOGS(_log, LOG_LVL_DEBUG,
+                         __func__ << " b booted uq=" << countQId.qId << ":count=" << countQId.count
+                                  << " tasksBooted=" << tasksBooted
+                                  << " uqBootedTasksSum=" << uqBootedTasksSum);
+                }
+            }
         }
     }
-    return false;
+    return;
 }
 
-/// @return true if all Tasks for this query are complete.
-/// Precondition, _qStatsMtx must be locked.
-bool QueryStatistics::_isMostlyDead() const { return _tasksCompleted >= _size; }
+bool QueriesAndChunks::_bootUserQuery(QueryStatistics::Ptr queryToBoot,
+                                      wsched::BlendScheduler::Ptr const& bSched) {
+    // Get query info
+    if (queryToBoot == nullptr) {
+        LOGS(_log, LOG_LVL_ERROR, __func__ << " queryToBoot is nullptr");
+        // This really should never happen.
```

Comment:
Another confusing statement. Please, consdier throwing exceptions for all those "should never happen".

Reply:
I'll change the comments.